File: india_compliance/audit_trail/client_scripts/accounts_settings.js
frappe.ui.form.on("Accounts Settings", {
    refresh(frm) {
        if (!frm.doc.enable_audit_trail) return;

        frm.set_df_property("enable_audit_trail", "read_only", 1);

        // this should never happen, but just in case
        if (frm.doc.delete_linked_ledger_entries) {
            frm.set_value("delete_linked_ledger_entries", 0);
        }

        toggle_linked_ledger_entry_deletion(frm);
    },
    enable_audit_trail(frm) {
        toggle_linked_ledger_entry_deletion(frm);
        if (!frm.doc.enable_audit_trail) return;

        let message = __("Audit Trail cannot be disabled once enabled.");
        if (frm.doc.delete_linked_ledger_entries) {
            message += __(
                `<br><br>
                Additionally, the following setting will be disabled
                to ensure Audit Trail integrity:<br>
                <strong>{0}</strong>`,
                [
                    __(
                        frappe.meta.get_label(
                            "Accounts Settings",
                            "delete_linked_ledger_entries",
                            frm.doc.name
                        )
                    ),
                ]
            );
            frm.set_value("delete_linked_ledger_entries", 0);
        }

        frappe.msgprint({
            title: __("Warning"),
            indicator: "orange",
            message,
        });
    },
});

function toggle_linked_ledger_entry_deletion(frm) {
    frm.set_df_property(
        "delete_linked_ledger_entries",
        "read_only",
        frm.doc.enable_audit_trail
    );
    frm.set_df_property(
        "delete_linked_ledger_entries",
        "description",
        frm.doc.enable_audit_trail
            ? "This setting has been disabled to ensure Audit Trail integrity."
            : ""
    );
}

File: india_compliance/audit_trail/overrides/accounts_settings.py
import frappe
from frappe import _

from india_compliance.audit_trail.setup import create_property_setters_for_versioning


def validate(doc, method=None):
    validate_change_in_enable_audit_trail(doc)
    validate_delete_linked_ledger_entries(doc)


def validate_change_in_enable_audit_trail(doc):
    if not doc.has_value_changed("enable_audit_trail"):
        return

    if not doc.enable_audit_trail:
        frappe.throw(_("Audit Trail cannot be disabled once enabled"))

    # Enable audit trail
    doc.delete_linked_ledger_entries = 0
    frappe.enqueue(create_property_setters_for_versioning, queue="short", at_front=True)


def validate_delete_linked_ledger_entries(doc):
    if doc.enable_audit_trail and doc.delete_linked_ledger_entries:
        frappe.throw(
            _("{0} cannot be enabled to ensure Audit Trail integrity").format(
                frappe.bold(doc.meta.get_label("delete_linked_ledger_entries"))
            )
        )
